runtime/src/iree/tooling/comparison.cc

```diff
@@ -0,0 +1,282 @@
+      case IREE_VM_VALUE_TYPE_I8:
+        type = "i8";
+        break;
+      case IREE_VM_VALUE_TYPE_I16:
+        type = "i16";
+        break;
+      case IREE_VM_VALUE_TYPE_I32:
+        type = "i32";
+        break;
+      case IREE_VM_VALUE_TYPE_I64:
+        type = "i64";
+        break;
+      case IREE_VM_VALUE_TYPE_F32:
+        type = "f32";
+        break;
+      case IREE_VM_VALUE_TYPE_F64:
+        type = "f64";
+        break;
+      default:
+        type = "?";
+        break;
+    }
+    return iree_string_builder_append_cstring(builder, type);
+  } else if (iree_vm_variant_is_ref(variant)) {
+    return iree_string_builder_append_string(
+        builder, iree_vm_ref_type_name(variant.type.ref_type));
+  } else {
+    return iree_string_builder_append_string(builder, IREE_SV("unknown"));
+  }
+}
+
+static bool iree_tooling_compare_values(int result_index,
+                                        iree_vm_variant_t expected_variant,
+                                        iree_vm_variant_t actual_variant,
+                                        iree_string_builder_t* builder) {
+  IREE_ASSERT_EQ(expected_variant.type.value_type,
+                 actual_variant.type.value_type);
+  switch (expected_variant.type.value_type) {
+    case IREE_VM_VALUE_TYPE_I8:
+      if (expected_variant.i8 != actual_variant.i8) {
+        IREE_CHECK_OK(iree_string_builder_append_format(
+            builder,
+            "[FAILED] result[%d]: i8 values differ\n  expected: %" PRIi8
+            "\n  actual: %" PRIi8 "\n",
+            result_index, expected_variant.i8, actual_variant.i8));
+        return false;
+      }
+      return true;
+    case IREE_VM_VALUE_TYPE_I16:
+      if (expected_variant.i16 != actual_variant.i16) {
+        IREE_CHECK_OK(iree_string_builder_append_format(
+            builder,
+            "[FAILED] result[%d]: i16 values differ\n  expected: %" PRIi16
+            "\n  actual: %" PRIi16 "\n",
+            result_index, expected_variant.i16, actual_variant.i16));
+        return false;
+      }
+      return true;
+    case IREE_VM_VALUE_TYPE_I32:
+      if (expected_variant.i32 != actual_variant.i32) {
+        IREE_CHECK_OK(iree_string_builder_append_format(
+            builder,
+            "[FAILED] result[%d]: i32 values differ\n  expected: %" PRIi32
+            "\n  actual: %" PRIi32 "\n",
+            result_index, expected_variant.i32, actual_variant.i32));
+        return false;
+      }
+      return true;
+    case IREE_VM_VALUE_TYPE_I64:
+      if (expected_variant.i64 != actual_variant.i64) {
+        IREE_CHECK_OK(iree_string_builder_append_format(
+            builder,
+            "[FAILED] result[%d]: i64 values differ\n  expected: %" PRIi64
+            "\n  actual: %" PRIi64 "\n",
+            result_index, expected_variant.i64, actual_variant.i64));
+        return false;
+      }
+      return true;
+    case IREE_VM_VALUE_TYPE_F32:
+      // TODO(benvanik): use tolerance flag.
```

Comment:
Can we please set a default tolerance? x86_64 result is different from riscv64 result, and I would like to use the same expected output for the test.

Reply:
a default tolerance of 0.001 should be sufficient.

Reply:
these are only used for scalar returns, which AFAIK no frontend actually produces yet